### DISCUSSION TITLE
Infallible BodyError guidance

### DISCUSSION THREAD

**Discussion by kellytk** [tokio-rs/axum#505] [Q&A] [answered]
When should `Infallible` be used for `IntoResponse`'s `BodyError` and why is it not the [default recommendation](https://docs.rs/axum/0.3.2/src/axum/response/mod.rs.html#173)?

**Comment by davidpdrsn** [comment, marked as answer]
The `BodyError` associated type only exists to make `impl IntoResponse` work, without it you would have to specify the exact body type. As such you cannot pick whatever error type you want. It must always match the `Body` associated type. For example if you're using `axum::body::Body` the error must be `hyper::Error` because that how the trait is implemented in [hyper](https://docs.rs/hyper/0.14.14/hyper/struct.Body.html#impl-Body). So you should always set it to

```rust
type BodyError = <Self::Body as axum::body::HttpBody>::Error;
```

Then rust can infer the error type from the body without you having to specify it.

Generally the error type will only be infallible if you're using [`Full`](https://docs.rs/http-body/0.4.4/http_body/struct.Full.html) or [`Empty`](https://docs.rs/http-body/0.4.4/http_body/struct.Empty.html) as the body type.

**Reply by kellytk** [reply]
Thank you for the helpful information.  The [documentation](https://docs.rs/axum/0.3.2/src/axum/response/mod.rs.html#156) note on `Full` states:

    When you know your response always contains exactly one chunk.

How can that fact be ascertained?

**Reply by davidpdrsn** [reply]
If your body is not empty and is not a stream, then you probably want full. For example sending some JSON requires having all the JSON ready upfront. Thus it can be sent in one chunk.

**Reply by jplatte** [reply]
To be clear, streaming serialization of JSON is absolutely *possible*, it's just not something `serde_json` supports (outside of sync `io::Writer`s).
